Encryption.md

```diff
@@ -19,216 +19,544 @@
+
+In the plaintext footer mode, the `optional ColumnMetaData meta_data` is set in the `ColumnChunk`
+structure for all columns, but is stripped of the statistics for the sensitive (encrypted) 
+columns. These statistics are available for new readers with the column key - they decrypt 
+the `encrypted_column_metadata` field, described in the section 5.1, and parse it to get statistics 
+and all other column metadata values. The legacy readers are not aware of the encrypted metadata field; 
+they parse the regular (plaintext) field as usual. While they can't read the data of the encrypted 
+columns, they read the metadata to extract the offset and size of the column data - required for 
+input vectorization (see the section 5.4).
+
+The plaintext footer can be signed with a footer key in order to prevent tampering with the 
+`FileMetaData` contents. The footer signing is done by encrypting the serialized `FileMetaData` 
+structure with the 
+AES GCM algorithm - using the footer key, and an AAD constructed according to the instructions 
+of the section 4.4. Only the nonce and GCM tag are stored in the file – as a 28-byte 
+fixed-length array, written right after  the footer itself. The ciphertext is not stored, 
+because it is not required for footer integrity verification by readers.
+
+| nonce (12 bytes) |  tag (16 bytes) |
+|------------------|-----------------|
+
+
+The plaintext footer mode sets the following fields in the the FileMetaData structure:
+
+```c
+struct FileMetaData {
+...
+  /** 
+   * Encryption algorithm. Note that this field is set only in encrypted files
+   * with plaintext footer. Files with encrypted footer store the algorithm id
+   * in FileCryptoMetaData structure.
+   */
+  8: optional EncryptionAlgorithm encryption_algorithm
+
+  /** 
+   * Set only for encrypted files with a plaintext footer. 
+   * true if footer signature is written in the file. 
+   */
+  9: optional bool signed_footer
+
+  /** 
+   * Retrieval metadata of key used for signing the footer. 
+   * Set only for encrypted files with a plaintext and signed footer. 
+   */ 
+  10: optional binary footer_signing_key_metadata
+}
+```
+ 
+The `FileMetaData` structure is Thrift-serialized and written to the output stream.
+
+The plaintext footer and an optional signature are followed by a 4-byte little endian integer 
+that contains either the footer length - or a combined length of the footer and its 28-byte 
+signature, if the footer is signed. A final magic string, "PAR1", is written at the end of the 
+file. The same magic string is written at the beginning of the file (offset 0). The magic bytes 
+for plaintext footer mode are ‘PAR1’ to allow legacy readers to read projections of the file 
+that do not include encrypted columns.
+
+<!-- 
+ ![File Layout - Encrypted footer](doc/images/FileLayoutEncryptionPF.jpg)
+   -->
+
+ 
+### 5.4 New fields for vectorized readers
```

Comment:
What broke that requires this?

I think this is an addition to the Parquet format, not to encryption. Why is it needed for encryption but not for normal vectorized reads? Can't total compressed size be inferred from page locations?

Also, `total_compressed_size` is misleading. It is total size on disk, not just the compressed size. Page buffers are compressed, but may also be encrypted. Page headers aren't compressed.

Reply:
This helps normal vectorized reads too - instead of looping on all column chunks, they can just read the row group size field.

The column chunk length is not available in files with encrypted footer (unless a reader has all column keys) - therefore, these new fields are required there.  It might be possible to reconstruct the size from page lengths and other available file metadata, but it wont be easy or elegant ( eg the row group offset is also unknown, so we wouldn't know where the pages start; its also not obvious where they end for each row group). 
I think adding these two fields is the right thing to do for both regular and encrypted files.

Regarding the name `total_compressed_size` - it copies an existing field in `ColumnMetaData` structure, called the same:
```c
/** total byte size of all compressed pages in this column chunk (including the headers) **/
  7: required i64 total_compressed_size
```

Reply:
regarding section 5.4 - we can replace it by adding a short text in section "5.2 Encrypted footer mode", explaining that these two fields must be set in RowGroup structures in files created in this mode.

Reply:
> I think adding these two fields is the right thing to do for both regular and encrypted files

If that's the case, then we should add the fields to the spec as a separate change.

A reader can find the starting point of all of the columns it intends to read and work from there. Vectorized reads don't necessarily need a stopping point and that can be inferred most of the time.

I don't see a need to over-complicate this spec with this concern, let's do it separately.

Reply:
Ok.